array.rb

```diff
@@ -1,4 +1,58 @@
 class Array
+  # call-seq:
+  #   array.each {|element| ... } -> self
+  #   array.each -> Enumerator
+  #
+  # Iterates over array elements.
+  #
+  # When a block given, passes each successive array element to the block;
+  # returns +self+:
+  #
+  #   a = [:foo, 'bar', 2]
+  #   a.each {|element|  puts "#{element.class} #{element}" }
+  #
+  # Output:
+  #
+  #   Symbol foo
+  #   String bar
+  #   Integer 2
+  #
+  # Allows the array to be modified during iteration:
+  #
+  #   a = [:foo, 'bar', 2]
+  #   a.each {|element| puts element; a.clear if element.to_s.start_with?('b') }
+  #
+  # Output:
+  #
+  #   foo
+  #   bar
+  #
+  # When no block given, returns a new Enumerator:
+  #   a = [:foo, 'bar', 2]
+  #
+  #   e = a.each
+  #   e # => #<Enumerator: [:foo, "bar", 2]:each>
+  #   a1 = e.each {|element|  puts "#{element.class} #{element}" }
+  #
+  # Output:
+  #
+  #   Symbol foo
+  #   String bar
+  #   Integer 2
+  #
+  # Related: #each_index, #reverse_each.
+  def each
+    unless block_given?
+      return to_enum(:each) { self.length }
+    end
```

Comment:
```suggestion
    unless defined?(yield)
      return Primitive.cexpr!(%q{ SIZED_ENUMERATOR(self, 0, 0, ary_enum_length) })
    end
```

Reply:
IMO it's cleaner to use the idiomatic plain Ruby code for this (i.e. before this suggestion).
It's also good because it encourages JITs and the interpreter to optimize patterns used in Ruby gems & apps.

As more of the core library becomes written in Ruby I think it's valuable it looks like regular Ruby code whenever possible vs patching pieces of C code together.

Reply:
I agree with Benoit.